### PR TITLE
Feat: #193 Dapp connector

### PR DESCRIPTION
**Pull request by hoailinh1210**
### Summary (Please recap what you changed or fixed)
- User can:
     - Connect your site
     - Disconnect your site
     - Sign deploy hash
     - Sign message
     - Event:
            - connected
            - disconnected
            - activeKeyChange

For testing purpose:
- Please change capserHelpers global: `window.casperlabsHelper = window.casperDashPluginHelpers` with console

### Checklist (If you won't pass this checklist please comment why)

-   [x] I removed all commented or unnecessary code.
-   [x] Provide the screenshots due to UI changed or bug fixed
-   [x] My code has covered these test cases (please list down your tested cases):

### CODE REVIEW

File: client/src/components/Common/MiddleTruncatedText/index.jsx [hunk 1]
```diff
@@ -4,6 +4,9 @@ import { getEndString } from '../../../helpers/format';
 import './MiddleTruncatedText.scss';
 
 export const MiddleTruncatedText = ({ children, end, placement = 'top' }) => {
+	if (!children) {
+		return null;
+	}
```
Comment (duc-q) [inline client/src/components/Common/MiddleTruncatedText/index.jsx:9]: This indicates something loosely validated from parent, also an not a best practice for minor component.
I think from the parent's component we should check for proper condition before rendering `<MiddleTruncatedText />`.

File: client/src/components/web-extension/ConnectedSites/index.jsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            const tabs = await chrome.tabs.query({
+                active: true,
+                lastFocusedWindow: true
+            });
+            if (_isEmpty(tabs)) {
+                return;
+            }
+            const [currentTab] = tabs;
+            const currentTabUrl = _get(currentTab, 'url', false);
+            if (!_get(currentTab, 'url', false)) {
```
Comment (duc-q) [inline client/src/components/web-extension/ConnectedSites/index.jsx:30]: This should be `if (!currentTabUrl)`, right?

File: client/src/components/web-extension/ConnectedSites/index.jsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                return;
+            }
+            const url = new URL(currentTabUrl);
+
+            const publicKeys = Object.keys(sites).reduce((curry, publicKey) => {
```
Comment (duc-q) [inline client/src/components/web-extension/ConnectedSites/index.jsx:35]: Better replace `curry` variable name with something like `result`. I don't see any curried function in this reduce.

File: client/src/components/web-extension/ConnectedSites/index.jsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            setPublicKeys(publicKeys);
+        }
+
+        getPublicKeys();
+    }, []);
+
+    const handleOnDisconnect = async (publicKey) => {
+        await disconnectFromSite(currentSite, publicKey);
+
+        setPublicKeys((publicKeys) => publicKeys.filter(key => key !== publicKey));
```
Comment (duc-q) [inline client/src/components/web-extension/ConnectedSites/index.jsx:57]: If using function style in `setState`, a good practice is having `prev` prefix. E.g: `prevPublicKeys`, this is much clearer

File: client/src/components/web-extension/ConnectedSites/index.jsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            </div>
+            <div className="public_keys">
+                {currentPublicKeys.map((publicKey) => {
+                    return (
+                        <div className="public_key_item" key={publicKey}>
+                            <div className="user_icon">
+                                <UserIcon/>
+                            </div>
+                            <div className="text">
+                                <MiddleTruncatedText end={4}>{publicKey}</MiddleTruncatedText>
```
Comment (duc-q) [inline client/src/components/web-extension/ConnectedSites/index.jsx:76]: This relates to my comment on `<MiddleTruncatedText />` . Is there any chance `publicKey` would be empty? If so, we can return `null` right from here

File: client/src/components/web-extension/DappSignMessageRequest/index.jsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import { approveSignMessageRequest, closePopup, parseMessageData } from '@cd/components/hooks/useServiceWorker';
+import { withServiceWorkerRequired } from '@cd/components/hocs/ServiceWorkerRequired';
+import { MiddleTruncatedText } from '@cd/components/Common/MiddleTruncatedText';
+
+import './index.scss';
+
+const DappSignMessageRequest = () => {
+	const [messageData, setMessageData] = useState({});
+
+	useEffect(() => {
```
Comment (duc-q) [inline client/src/components/web-extension/DappSignMessageRequest/index.jsx:12]: Move `useEffect` as close to `return` as possible

File: client/src/content/content.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+function injectCustomJs() {
+    try {
+      let jsPath = '/static/js/scripts/content/inpage.js';
+      const container = document.head || document.documentElement;
+      const scriptTag = document.createElement('script');
+      scriptTag.setAttribute('type', 'text/javascript');
+      scriptTag.src = chrome.runtime.getURL(jsPath);
+      container.insertBefore(scriptTag, container.children[0]);
+      scriptTag.onload = function () {
+        // remove after run the script
```
Comment (duc-q) [inline client/src/content/content.js:12]: Can you please add more comment here to explain why we need to remove the script after executing it? Thanks

File: client/src/services/ServiceWorker/Controllers/PopupController.js [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    getConnectedSites = async () => {
+        const valueObj = await getChromeStorageLocal(CONNECTED_SITES);
+
+        return _get(valueObj, CONNECTED_SITES, {});
+    }
+
+    addConnectedSite = async ({ site, publicKey }) => {
+        let connectedSites = await this.getConnectedSites();
+        const sites = _get(connectedSites, publicKey, false);
```
Comment (duc-q) [inline client/src/services/ServiceWorker/Controllers/PopupController.js:188]: Minor: Should not mix variable type. I'm assuming the default value here would be `undefined` or empty array `[ ]`.

File: client/src/services/ServiceWorker/Controllers/SigningController.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
+import EventEmitter from 'events';
+import {
+    DeployUtil,
+    encodeBase16,
+    formatMessageWithHeaders,
+} from 'casper-js-sdk';
+import _pick from 'lodash-es/pick';
+import { getDeployArgs, getDeployPayment, getDeployType } from '../utils';
```
Comment (duc-q) [inline client/src/services/ServiceWorker/Controllers/SigningController.js:8]: Convert to alias path

File: client/src/services/ServiceWorker/Controllers/SigningController.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        let innerDeploy = DeployUtil.deployFromJson(deployJson.deploy);
+        if (!innerDeploy.ok) {
+            innerDeploy.mapErr(err => {
+                throw new Error(err.message);
+            });
+        }
+
+        this.unsignedDeploy = {
+            status: SIGNING_STATUS.unsigned,
+            deploy: innerDeploy.unwrap(),
```
Comment (duc-q) [inline client/src/services/ServiceWorker/Controllers/SigningController.js:58]: Can you please comment why we need `unwrap` here? Thanks.
Ref: 
https://casper-ecosystem.github.io/casper-js-sdk/latest/modules/_lib_deployutil_.html#deployfromjson

**Review by duc-q — COMMENTED** [review]
One more minor thing, should we use `dAppCamelCase` instead of `dappCamelCase` ?

File: client/src/services/ServiceWorker/Controllers/SigningController.js (hunk 10 above)
Comment (hoailinh1210) [reply client/src/services/ServiceWorker/Controllers/SigningController.js:58]: We will get the deploy hash when using unwrap function

File: client/src/components/Common/MiddleTruncatedText/index.jsx (hunk 1 above)
Comment (hoailinh1210) [reply client/src/components/Common/MiddleTruncatedText/index.jsx]: Just remove this

File: client/src/content/content.js (hunk 7 above)
Comment (hoailinh1210) [reply client/src/content/content.js]: Hello @duc-q bro, please take a look, I commented on it

File: client/src/components/web-extension/DappConnection/index.jsx [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		try {
+			await addConnectedSite(connectedUrl, publicKey);
+		} catch(err) {
+			// eslint-disable-next-line no-console
+			console.log(err);
+		}
+	}
+
+	const onCancel = () => {
+		closePopup();
```
Comment (djhahe) [inline client/src/components/web-extension/DappConnection/index.jsx:35]: need to send message back to dapp

File: client/src/components/web-extension/DappSignDeployRequest/index.jsx [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		try {
+			await approveSignDeployRequest();
+		} catch (err) {
+			// eslint-disable-next-line no-console
+			console.log(err);
+		}
+	};
+
+	const onCancel = () => {
+		closePopup();
```
Comment (djhahe) [inline client/src/components/web-extension/DappSignDeployRequest/index.jsx:33]: need to send message to dapp

**Review by djhahe — CHANGES_REQUESTED** [review]

File: client/src/components/web-extension/DappSignDeployRequest/index.jsx (hunk 12 above)
Comment (hoailinh1210) [reply client/src/components/web-extension/DappSignDeployRequest/index.jsx]: Just updated

File: client/src/components/web-extension/ConnectedSites/index.jsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+                return;
+            }
+            const [currentTab] = tabs;
+            const currentTabUrl = _get(currentTab, 'url', false);
+            if (!currentTabUrl) {
+                return;
+            }
+            const url = new URL(currentTabUrl);
+
+            const publicKeys = Object.keys(sites).reduce((publicKeys, publicKey) => {
```
[the remark this answers is not in the document]
Comment (hoailinh1210) [reply client/src/components/web-extension/ConnectedSites/index.jsx]: Thanks for your suggestion, I just improve on it with following code:
```
const publicKeys = Object.keys(sites).filter(publicKey => {
          return _get(sites, publicKey, []).includes(url.origin);
});
```

File: client/src/components/Common/Auth/useWithAccount.js [hunk 14]
```diff
@@ -6,22 +6,41 @@ import { getConnectedAccountChromeLocalStorage } from '@cd/actions/userActions.u
...
+
 		/**
 		 * Make sure user must be valid before navigating to other route
 		 */
 		const isValidUserShape = Boolean(user && !isEmpty(user?.loginOptions));
 		if (isValidUserShape) {
-			navigate(!user.publicKey ? '/welcomeBack' : '/');
-			return;
+			switch(true) {
+				case pathname === '/dappConnect':
```
[the remark this answers is not in the document]
Comment (hoailinh1210) [reply client/src/components/Common/Auth/useWithAccount.js:24]: I will create another PR for it

[the remark this answers is not in the document]
Comment (hoailinh1210) [reply client/src/components/Common/Auth/useWithAccount.js:24]: Hi @djhahe,
I created a new PR for this,
https://github.com/CasperDash/casperdash-client/pull/201
Please take a look.

**Review by djhahe — APPROVED** [review]

**Pull request merged**